### PR TITLE
have verify_account_authority check passed-in keys

### PR DESCRIPTION
**Pull request by jmjatlanta**
Potential fix for #1366 

The original code always returned false, due to not using the passed in keys to do anything.

This PR uses a different approach. Compare the public keys of what was passed in versus what is known.

### CODE REVIEW

File: libraries/app/database_api.cpp [hunk 1]
```diff
@@ -167,7 +167,7 @@ class database_api_impl : public std::enable_shared_from_this<database_api_impl>
       vector<withdraw_permission_object> get_withdraw_permissions_by_giver(const std::string account_id_or_name, withdraw_permission_id_type start, uint32_t limit)const;
       vector<withdraw_permission_object> get_withdraw_permissions_by_recipient(const std::string account_id_or_name, withdraw_permission_id_type start, uint32_t limit)const;
 
-   //private:
+   // private:
```
Comment (oxarbitrage) [inline libraries/app/database_api.cpp]: we maybe just remove this comment ?

Comment (jmjatlanta) [reply libraries/app/database_api.cpp]: Yes, I tried (hence the space I accidentally added). It did not compile due to _db being used outside, plus some others. I was thinking of fixing all of it, but resisted the urge because I was unsure why it was originally commented out.

**Comment by abitmore** [comment]
IMHO this is not the correct approach, since you ignored multi-auth accounts at all.

Correct approach would be to call `verify_authority` function somehow with available public keys as "approved".

**Comment by jmjatlanta** [comment]
Thank you @abitmore for partially answering a question I had. 

Is the purpose of this function to prove that the public keys are valid? Or is this function saying "If you send a transaction with the private keys related to these public keys, it will go through"?

If you want to prove that those public keys are related to this account, well, I think I am done.

If you want to take in to account multi-auth accounts, then more will need to be added. And now with projects being worked on that make keys more granular, I believe it would depend on what type of transaction you are doing as to if such a signature would pass or fail.

If my thinking and assumptions on the last paragraph are correct, perhaps it would be best to pause working on this issue until those features are added.

Thoughts?

**Comment by pmconrad** [comment]
Doc comment says
> @return true if the signers have enough authority to authorize an account

IMO it is clear that multi-level signatures must be taken into account.

File: libraries/app/database_api.cpp [hunk 2]
```diff
@@ -2070,29 +2070,27 @@ bool database_api::verify_account_authority( const string& account_name_or_id, c
...
       const flat_set<public_key_type>& keys )const
 {
-   const account_object* account = get_account_from_string(account_name_or_id);
+   // create a dummy transfer
+   transfer_operation op;
+   op.from = get_account_from_string(account_name_or_id)->id;
+   std::vector<operation> ops;
+   ops.emplace_back(op);
+
+   try
```
Comment (cogutvalera) [inline libraries/app/database_api.cpp]: why not just to create `signed_transaction` and call it's `verify_authority` method ?
So the full implementation of `verify_account_authority` might be like next:
```
bool database_api_impl::verify_account_authority( const string& account_name_or_id, 
      const flat_set<public_key_type>& keys )const
{
   // create a dummy transfer
   transfer_operation op;
   op.from = get_account_from_string(account_name_or_id)->id;
   std::vector<operation> ops;
   ops.emplace_back(op);
   signed_transaction trx;
   trx.signees = keys;
   trx.operations.push_back(op);

   return verify_authority(trx);
}
```
It's less code and more simple IMHO.

Comment (pmconrad) [reply libraries/app/database_api.cpp]: Nice idea, but IMO the signees field should be used as an internal cache only, not part of the tx struct's interface.

Comment (cogutvalera) [reply libraries/app/database_api.cpp]: `graphene::chain::verify_authority` will use `signees` returned by `signed_transaction::get_signature_keys` method, so it should work

**Comment by jmjatlanta** [comment]
I have added a muti-signature test (2 of 3) to wallet_test. Please let me know what you think.

File: tests/tests/wallet_tests.cpp [hunk 3]
```diff
@@ -95,4 +96,54 @@ BOOST_FIXTURE_TEST_SUITE(wallet_tests, database_fixture)
...
+      fc::ecc::private_key nathan_key2 = fc::ecc::private_key::regenerate(fc::digest("key2"));
+      fc::ecc::private_key nathan_key3 = fc::ecc::private_key::regenerate(fc::digest("key3"));
+      const account_object& nathan = create_account("nathan", nathan_key1.get_public_key() );
+      fund(nathan);
+      graphene::app::database_api db_api(db);
+
+      try {
+         account_update_operation op;
+         op.account = nathan.id;
+         op.active = authority(2, public_key_type(nathan_key1.get_public_key()), 1, public_key_type(nathan_key2.get_public_key()), 1, public_key_type(nathan_key3.get_public_key()), 1);
```
Comment (cogutvalera) [inline tests/tests/wallet_tests.cpp]: this line is too long, please reduce it something like next line of code:
```
         op.active = authority(
            2, public_key_type(nathan_key1.get_public_key()), 
            1, public_key_type(nathan_key2.get_public_key()), 
            1, public_key_type(nathan_key3.get_public_key()), 
            1
         );
```

**Comment by cogutvalera** [comment]
All tests (`verify_account_authority` & `any_two_of_three`) have passed successfully with no errors.
Nice work ! Thanks !

**Comment by cogutvalera** [comment]
IMHO just required little improvements with that line as commented above ...

File: tests/tests/wallet_tests.cpp [hunk 4]
```diff
@@ -70,10 +71,80 @@ BOOST_FIXTURE_TEST_SUITE(wallet_tests, database_fixture)
...
+      {
+     	   flat_set<public_key_type> public_keys;
+         public_keys.emplace(nathan_key1.get_public_key());
+     	   BOOST_CHECK(!db_api.verify_account_authority("nathan", public_keys));
+      }
+   } catch (fc::exception& e) {
+      edump((e.to_detail_string()));
+      throw;
+   }
+}
```
Comment (pmconrad) [inline tests/tests/wallet_tests.cpp]: Please add a test where nathan's account has a multisig from one key and two other accounts.

Comment (jmjatlanta) [reply tests/tests/wallet_tests.cpp]: Added what I think you are asking. It is not much different than the last, so I may be misunderstanding your request.

File: tests/tests/wallet_tests.cpp [hunk 5]
```diff
@@ -70,10 +71,122 @@ BOOST_FIXTURE_TEST_SUITE(wallet_tests, database_fixture)
...
+      create_account("alice", alice_key.get_public_key() );
+      create_account("bob", bob_key.get_public_key() );
+
+      graphene::app::database_api db_api(db);
+
+      try {
+         account_update_operation op;
+         op.account = nathan.id;
+         op.active = authority(3, public_key_type(nathan_key.get_public_key()), 1, 
+               public_key_type(alice_key.get_public_key()), 1, public_key_type(bob_key.get_public_key()), 1);
```
Comment (pmconrad) [inline tests/tests/wallet_tests.cpp]: Here you should use alice_id and bob_id, not their keys

File: tests/tests/wallet_tests.cpp [hunk 6]
```diff
@@ -70,10 +71,122 @@ BOOST_FIXTURE_TEST_SUITE(wallet_tests, database_fixture)
...
+      throw;
+   }
+}
+
+BOOST_AUTO_TEST_CASE( verify_authority_multiple_accounts )
+{
+   try {
+      fc::ecc::private_key nathan_key = fc::ecc::private_key::regenerate(fc::digest("key1"));
+      fc::ecc::private_key alice_key = fc::ecc::private_key::regenerate(fc::digest("key2"));
+      fc::ecc::private_key bob_key = fc::ecc::private_key::regenerate(fc::digest("key3"));
```
Comment (pmconrad) [inline tests/tests/wallet_tests.cpp]: Please use ``ACTORS`` macro

File: tests/tests/wallet_tests.cpp [hunk 7]
```diff
@@ -70,10 +71,122 @@ BOOST_FIXTURE_TEST_SUITE(wallet_tests, database_fixture)
...
+
+      // requires 3 signatures
+      {
+      	flat_set<public_key_type> public_keys;
+      	public_keys.emplace(nathan_key.get_public_key());
+      	public_keys.emplace(alice_key.get_public_key());
+      	public_keys.emplace(bob_key.get_public_key());
+      	BOOST_CHECK(db_api.verify_account_authority("nathan", public_keys));
+      }
+
```
Comment (pmconrad) [inline tests/tests/wallet_tests.cpp]: Also need a negative test, e. g. alice_key missing.

File: tests/tests/wallet_tests.cpp (hunk 4 above)
Comment (cogutvalera) [reply tests/tests/wallet_tests.cpp]: IMHO we should check different combinations as in `any_two_of_three` test, perhaps the same combinations, but with 3 accounts, 1 key per 1 account, just my thoughts ...

File: tests/tests/wallet_tests.cpp [hunk 8]
```diff
@@ -149,39 +149,38 @@ BOOST_AUTO_TEST_CASE( any_two_of_three )
...
 
       graphene::app::database_api db_api(db);
 
       try {
          account_update_operation op;
          op.account = nathan.id;
-         op.active = authority(3, public_key_type(nathan_key.get_public_key()), 1, 
-               public_key_type(alice_key.get_public_key()), 1, public_key_type(bob_key.get_public_key()), 1);
+         op.active = authority(3, nathan_public_key, 1, 
+               alice.id, 1, bob.id, 1);
```
Comment (cogutvalera) [inline tests/tests/wallet_tests.cpp]: IMHO we don't need to split this line, it isn't long now

**Comment by pmconrad** [comment]
Good, thanks!

File: libraries/app/include/graphene/app/database_api.hpp [hunk 9]
```diff
@@ -663,9 +663,12 @@ class database_api
       bool           verify_authority( const signed_transaction& trx )const;
 
       /**
-       * @return true if the signers have enough authority to authorize an account
+       * @brief Verify that the public keys have enough authority to authorize a transaction
```
Comment (abitmore) [inline libraries/app/include/graphene/app/database_api.hpp]: Why this change? There is no transaction in the API.

File: libraries/app/include/graphene/app/database_api.hpp [hunk 10]
```diff
@@ -663,9 +663,12 @@ class database_api
...
+       * @param account_name_or_id the account to check
+       * @param signers the public keys
+       * @return true if the passed in keys have enough authority to authorize a transaction
```
Comment (abitmore) [inline libraries/app/include/graphene/app/database_api.hpp]: Why this? There is no transaction in the API.

File: tests/tests/wallet_tests.cpp [hunk 11]
```diff
@@ -70,10 +71,120 @@ BOOST_FIXTURE_TEST_SUITE(wallet_tests, database_fixture)
...
           }
 
       } FC_LOG_AND_RETHROW()
   }
 
+  BOOST_AUTO_TEST_CASE(verify_account_authority) {
+      try {
+         
+         ACTORS( (nathan) );
+         graphene::app::database_api db_api(db);
```
Comment (abitmore) [inline tests/tests/wallet_tests.cpp]: I think it's better to move these test cases to database_api_tests.cpp.

**Comment by abitmore** [comment]
Code looks fine to me.

File: libraries/app/include/graphene/app/database_api.hpp [hunk 12]
```diff
@@ -663,9 +663,12 @@ class database_api
       bool           verify_authority( const signed_transaction& trx )const;
 
       /**
-       * @return true if the signers have enough authority to authorize an account
+       * @brief Verify that the public keys have enough authority to approve an operation
```
Comment (abitmore) [inline libraries/app/include/graphene/app/database_api.hpp]: I think we should mention the account in this sentence and the return

Comment (jmjatlanta) [reply libraries/app/include/graphene/app/database_api.hpp]: I think the latest commit will be more to your liking. Please take a look.
